index.js

```diff
@@ -1163,8 +1301,8 @@
       $pairs.push ([x, y]);
       try {
         return Z.Setoid.test (x) ?
-          Z.Setoid.methods.equals (x) (y) :
-          Object$prototype$equals.call (x, y);
+               Z.Setoid.methods.equals (x, y) :
+               Object$prototype$equals.call (x, y);
```

Comment:
There's an inefficiency in this algorithm which in theory can lead to large slow-downs for nested structures:

- Setoid.test dispatches to `prototypeMethod` to figure out if we're dealing with a setoid.
- `prototypeMethod` recursively calls `Setoid.test` for values in Arrays and Objects
- Once Setoid instance is recursively established, `Setoid.methods.equals` is called, which dispatches to `prototypeMethod` again.

Reply:
I think we should explore this issue separately from this PR.

Reply:
I have a new branch where I've started fixing these issues, but not without breaking some things. I'll open a PR for discussion.

Reply:
https://github.com/sanctuary-js/sanctuary-type-classes/pull/163